### PR TITLE
#1: General migration. All unit tests green.

### LINKED ISSUE #1: Extract BucketFS code from exasol-testcontainers
The BucketFS implementation is also required for other execution models for example a cluster on AWS.

Right now for such use cases we have to reimplement the uploading and checking code agian and again.

It would be very helpful to extract the BuckeFS code into a separate module and add a interface for accessing the log file.

One could go even further and extract a ExasolTestInterface with the public API of this packet and make it independent of the TestBackend (Testcontainers / local instance started via docker-compose / cluster on AWS / cluster on Azure / ...)

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #1.

After this initial migration, system spec and design need to be added and requirement tracing to be fixed. That will be done in a separate ticket.

### CODE REVIEW

File: src/main/java/com/exasol/bucketfs/ClusterConfigurationBucketFactory.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ */
+public final class ClusterConfigurationBucketFactory implements BucketFactory {
+    private final Map<String, WriteEnabledBucket> bucketsCache = new HashMap<>();
+    private final String ipAddress;
+    private final BucketFsSerivceConfigurationProvider serviceConfigurationProvider;
+    private final Map<Integer, Integer> portMappings;
+    private final BucketFsMonitor monitor;
+
+    /**
+     * Create a new instance of a BucketFactory.
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/bucketfs/ClusterConfigurationBucketFactory.java:20]:
```suggestion
     * Create a new instance of a {@link ClusterConfigurationBucketFactory}.
```

File: src/test/java/com/exasol/bucketfs/LogBasedBucketFsMonitor.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        this.detectorFactory = detectorFactory;
+    }
+
+    @Override
+    public boolean isObjectSynchronized(final ReadOnlyBucket bucket, final String pathInBucket, final Instant afterUTC)
+            throws InterruptedException, BucketAccessException {
+        try {
+            return createBucketLogPatternDetector(pathInBucket).isPatternPresentAfter(afterUTC);
+        } catch (
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/bucketfs/LogBasedBucketFsMonitor.java:25]:
```suggestion
```

File: src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    void testListBucketContents(final String pathInBucket) throws BucketAccessException, InterruptedException {
+        assertThat(getDefaultBucket().listContents(pathInBucket), hasItem(startsWith("ScriptLanguages")));
+    }
+
+    void testListBucketContentsOfIllegalPathThrowsException() {
+        assertThrows(BucketAccessException.class, () -> getDefaultBucket().listContents("illegal\\path"));
+    }
+
+    @Test
+    void testDownloadFileThrowsExceptionOnIllegalPathInBucket(@TempDir final Path tempDir) {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/bucketfs/ReadEnabledBucketIT.java:60]: it seems there's a test missing for a successful download

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        final Bucket bucket = getDefaultBucket();
+        final String content = "read me";
+        bucket.uploadStringContent(content, fileName);
+        final Path pathToFile = tempDir.resolve(fileName);
+        bucket.downloadFile(fileName, pathToFile);
+        assertThat(Files.readString(pathToFile), equalTo(content));
+    }
+
+    @Test
+    void testDownloadFileThrowsExceptionOnIllegalLocalPath(@TempDir final Path tempDir)
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:125]: seems like this test should be moved to `ReadEnabledBucketIT`

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    @Test
+    void testUploadContentToIllegalUrlThrowsException() {
+        assertThrows(BucketAccessException.class,
+                () -> getDefaultBucket().uploadStringContent("irrelevant content", "this\\is\\an\\illegal\\URL"));
+    }
+
+    // [itest->dsn~downloading-a-file-from-a-bucket~1]
+    @Test
+    void testDownloadFile(@TempDir final Path tempDir)
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:113]: seems like this test should be moved to `ReadEnabledBucketIT`

File: src/main/java/com/exasol/bucketfs/BucketFsMonitor.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+     * Check if the object with the given path is already synchronized.
+     *
+     * @param bucket       bucket
+     * @param pathInBucket path to the object located in the bucket
+     * @param afterUTC     point in UTC time after which the object synchronization counts
+     * @return {@code true} if the object exists in the bucket and is synchronized
+     * @throws InterruptedException  if an interrupt occurs while the object synchronization is being evaluated
+     * @throws BucketAccessException if the object in the bucket is inaccessible
+     */
+    public boolean isObjectSynchronized(final ReadOnlyBucket bucket, final String pathInBucket, final Instant afterUTC)
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/bucketfs/BucketFsMonitor.java:19]: is the `Instant afterUTC` parameter really needed?

What happens if an object is synchronized before this time? from the user perspective this is not OK?

**Review by redcatbear — COMMENTED** [review]
Answers.

File: doc/system_requirements.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+* BucketFS: service provided by the Exasol database that allows keeping files distributed across all data nodes of an Exasol cluster.
+
+## Features
+
+Features are the highest level requirements in this document that describe the main functionality of BucketFS Java.
+
+### BucketFS Access
+`feat~bucketfs-access~1`
+
+ETC provides access to the BucketFS service(s) of the Exasol database.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:43]:
```suggestion
BucketFS Java provides access to the BucketFS service(s) of the Exasol database.
```

File: doc/system_requirements.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+## Functional Requirements
+
+This section lists functional requirements from the user's perspective. The requirements are grouped by feature where they belong to a single feature.
+
+### BucketFS Access
+
+#### Bucket Content Listing
+`req~bucket-content-listing~1`
+
+ETC lists the contents of a bucket in BucketFS.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:56]:
```suggestion
BucketFS Java lists the contents of a bucket in BucketFS.
```

File: doc/system_requirements.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+Covers:
+
+* [feat~bucketfs-access~1](#bucketfs-access)
+
+Needs: dsn
+
+#### Uploading a File to BucketFS
+`req~uploading-a-file-to-bucketfs~1`
+
+ETC uploads a file from a locally accessible filesystem to a bucket.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:71]:
```suggestion
BucketFS Java uploads a file from a locally accessible filesystem to a bucket.
```

File: doc/system_requirements.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+Covers:
+
+* [feat~bucketfs-access~1](#bucketfs-access)
+
+Needs: dsn
+
+#### Uploading Text to a File in BucketFS
+`req~uploading-text-to-a-file-in-bucketfs~1`
+
+ETC uploads text (aka. a "string") to a file inside a bucket.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:86]:
```suggestion
BucketFS Java uploads text (aka. a "string") to a file inside a bucket.
```

File: doc/system_requirements.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+Covers:
+
+* [feat~bucketfs-access~1](#bucketfs-access)
+
+Needs: dsn
+
+#### Uploading InputStream to a File in BucketFS
+`req~uploading-input-stream-to-a-file-in-bucketfs~1`
+
+ETC uploads the contents of an InputStream to a file inside a bucket.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:101]:
```suggestion
BucketFS Java uploads the contents of an InputStream to a file inside a bucket.
```

File: doc/system_requirements.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+Covers:
+
+* [feat~bucketfs-access~1](#bucketfs-access)
+
+Needs: dsn
+
+#### Downloading a File from BucketFS
+`req~downloading-a-file-from-bucketfs~1`
+
+ETC downloads a file from a bucket to a locally accessible filesystem.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:116]:
```suggestion
BucketFS Java downloads a file from a bucket to a locally accessible filesystem.
```

File: doc/system_requirements.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+Covers:
+
+* [feat~bucketfs-access~1](#bucketfs-access)
+
+Needs: dsn
+
+#### Bucket Authentication
+`req~bucket-authentication~1`
+
+ETC handles the authentication against the buckets automatically.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:131]:
```suggestion
BucketFS Java handles the authentication against the buckets automatically.
```

File: doc/system_requirements.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+* [feat~bucketfs-access~1](#bucketfs-access)
+
+
+Needs: dsn
+
+#### Waiting for Bucket Content Synchronization
+`req~waiting-for-bucket-content-synchronization~1`
+
+ETC allows integrators to wait for bucket contents to be synchronized on a single node after uploading a file.
```
Comment (chiaradiamarcelo) [inline doc/system_requirements.md:147]:
```suggestion
BucketFS Java allows integrators to wait for bucket contents to be synchronized on a single node after uploading a file.
```

**Review by chiaradiamarcelo — DISMISSED** [review]

**Review by chiaradiamarcelo — APPROVED** [review]

**Pull request merged**